fluent/migrate/transforms.py

```diff
@@ -92,42 +85,85 @@ def __call__(self, ctx):
 
     @staticmethod
     def flatten_elements(elements):
-        '''Flatten a list of FTL nodes into valid Pattern's elements'''
-        flattened = []
+        '''Flatten a list of FTL nodes into an iterator over PatternElements.'''
         for element in elements:
             if isinstance(element, FTL.Pattern):
-                flattened.extend(element.elements)
+                # PY3 yield from element.elements
+                for child in element.elements:
+                    yield child
             elif isinstance(element, FTL.PatternElement):
-                flattened.append(element)
+                yield element
             elif isinstance(element, FTL.Expression):
-                flattened.append(FTL.Placeable(element))
+                yield FTL.Placeable(element)
             else:
                 raise RuntimeError(
                     'Expected Pattern, PatternElement or Expression')
-        return flattened
 
     @staticmethod
-    def prune_text_elements(elements):
-        '''Join adjacent TextElements and remove empty ones'''
-        pruned = []
-        # Group elements in contiguous sequences of the same type.
-        for elem_type, elems in itertools.groupby(elements, key=type):
-            if elem_type is FTL.TextElement:
-                # Join adjacent TextElements.
-                text = FTL.TextElement(''.join(elem.value for elem in elems))
-                # And remove empty ones.
-                if len(text.value) > 0:
-                    pruned.append(text)
+    def join_adjacent_elements(elements):
```

Comment:
I'm OK with doing this but I think this is opposite to the direction the fix to handle leading/trailing whitespace would want to go. Do you want to deprioritize or maybe even wontfix [bug 1374246](https://bugzilla.mozilla.org/show_bug.cgi?id=1374246)?

Reply:
Wouldn't leading/trailing just be trailing pass in `pattern_of`, cutting leading and trailing whitespace off of TextElement into a Placeable, or prefixing with an empty placeable?

Reply:
Yes, but it also needs more handling in `CONCAT` so that `CONCAT(COPY(), COPY())` doesn't produce things like `Foo{" "}Bar` when the first `COPY` has trailing whitespace (or the second one has leading whitespace). I'd like to tackle it in a separate PR.

I'll implement your original suggestion to join `TextElements` and `StringExpressions` alike. This might result in `TextElement("      Foo")` in some rare cases. The serializer currently serializes this with the whitespace (although I'd like to change the behavior in the future and throw: [bug 1397233](https://bugzilla.mozilla.org/show_bug.cgi?id=1397233)) but the parses ignores it, so in the end, the whitespace is lost.

This is actually the status quo for translations with leading or trailing whitespace, like `key = \u0020Foo`. I'll unskip two tests to make it explicit. Implementing your suggestion now will prepare us to solve the leading/trailing whitespace issue uniformly in all transforms, regardless of whether it was part of the legacy translation or got concatenated from another legacy translation.